libmisc/salt.c

```diff
@@ -12,19 +12,321 @@
+
+#ifdef USE_YESCRYPT
+/*
+ * Return a salt prefix specifying the cost for the YESCRYPT method.
+ */
+static /*@observer@*/const unsigned long YESCRYPT_salt_cost (/*@null@*/long *prefered_cost)
+{
+	unsigned long cost;
+
+	if (NULL == prefered_cost) {
+		cost = (unsigned long) getdef_num ("YESCRYPT_COST_FACTOR", Y_COST_DEFAULT);
+	} else {
+		cost = (unsigned long) *prefered_cost;
+	}
+
+	if (cost < Y_COST_MIN) {
+		cost = Y_COST_MIN;
+	}
+	if (cost > Y_COST_MAX) {
+		cost = Y_COST_MAX;
+	}
+
+	return cost;
+}
+#endif /* USE_YESCRYPT */
+
+/* Maximum size of the setting for crypt_gensalt(). */
+#define SETTING_SIZE 16
+
+/*
+ * Generate 8 base64 ASCII characters of random salt.  If MD5_CRYPT_ENAB
+ * in /etc/login.defs is "yes", the salt string will be prefixed by "$1$"
+ * (magic) and pw_encrypt() will execute the MD5-based FreeBSD-compatible
+ * version of crypt() instead of the standard one.
+ * Other methods can be set with ENCRYPT_METHOD
+ *
+ * The method can be forced with the meth parameter.
+ * If NULL, the method will be defined according to the MD5_CRYPT_ENAB and
+ * ENCRYPT_METHOD login.defs variables.
+ *
+ * If meth is specified, an additional parameter can be provided.
+ *  * For the SHA256 and SHA512 method, this specifies the number of rounds
+ *    (if not NULL).
+ *  * For the YESCRYPT method, this specifies the cost factor (if not NULL).
+ */
+/*@observer@*/const char *crypt_make_salt (/*@null@*//*@observer@*/const char *meth, /*@null@*/void *arg)
+{
+	const char *method;
+	char *retval = NULL;
+	char setting[SETTING_SIZE];
+        unsigned long rounds = 0;
+	static char result[CRYPT_GENSALT_OUTPUT_SIZE];
+
+	if (NULL != meth)
+		method = meth;
+	else {
+		method = getdef_str ("ENCRYPT_METHOD");
+		if (NULL == method) {
+			method = getdef_bool ("MD5_CRYPT_ENAB") ? "MD5" : "DES";
+		}
+	}
+
+	if (0 == strcmp (method, "MD5")) {
+		MAGNUM(setting, '1');
+#ifdef USE_BCRYPT
+	} else if (0 == strcmp (method, "BCRYPT")) {
+		BCRYPTMAGNUM(setting);
+		rounds = BCRYPT_salt_rounds((int *)arg);
+#endif /* USE_BCRYPT */
+#ifdef USE_YESCRYPT
+	} else if (0 == strcmp (method, "YESCRYPT")) {
+		MAGNUM(setting, 'y');
+		rounds = YESCRYPT_salt_cost((long *)arg);
+#endif /* USE_YESCRYPT */
+#ifdef USE_SHA_CRYPT
+	} else if (0 == strcmp (method, "SHA256")) {
+		MAGNUM(setting, '5');
+		rounds = SHA_salt_rounds((int *)arg);
+	} else if (0 == strcmp (method, "SHA512")) {
+		MAGNUM(setting, '6');
+		rounds = SHA_salt_rounds((int *)arg);
+#endif /* USE_SHA_CRYPT */
+	} else if (0 != strcmp (method, "DES")) {
+		for (int i = 0; i < SETTING_SIZE; ++i) {
+			setting[i] = '.';
+		}
+		setting[SETTING_SIZE - 1] = '\0';
+		fprintf (shadow_logfd,
+			 _("Invalid ENCRYPT_METHOD value: '%s'.\n"
+			   "Defaulting to DES.\n"),
+			 method);
+	}
+
+	retval = crypt_gensalt_rn(setting, rounds, NULL, 0,
+				  result, CRYPT_GENSALT_OUTPUT_SIZE);
+
+	if (retval == NULL) {
+		for (int i = 0; i < CRYPT_GENSALT_OUTPUT_SIZE; ++i) {
+			result[i] = '\0';
+		}
+	}
```

Comment:
Why not call `crypt_gensalt_rn` from inside the `if`, so that we wouldn't need to introduce a variable? And why not use `memset` (if zeroizing the entire buffer is somehow needed, which I doubt)?

Reply:
Fixed in rebased commit.